{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import kipoi\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import pickle\n",
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.data]\u001b[0m successfully loaded the dataloader 5UTR_Model/. from /data/ouga04b/ag_gagneur/home/karollus/5UTRModel/Collab/kipoi/5UTR_Model/dataloader.py::StrandedSequenceVariantDataloader\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.model]\u001b[0m Downloading model arguments weights from https://zenodo.org/record/3374833/files/utr_model_avgpool.h5?download=1\u001b[0m\n",
      "Using downloaded and verified file: /data/ouga04b/ag_gagneur/home/karollus/5UTRModel/Collab/kipoi/5UTR_Model/downloaded/model_files/weights/1bac703c830d6322dbace78721b1e405\n",
      "Using TensorFlow backend.\n",
      "2019-08-26 17:01:06.000173: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA\n",
      "2019-08-26 17:01:06.009370: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2095185000 Hz\n",
      "2019-08-26 17:01:06.011560: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x55f8e12ba180 executing computations on platform Host. Devices:\n",
      "2019-08-26 17:01:06.011597: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>\n",
      "2019-08-26 17:01:07.539427: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x55f8e133f190 executing computations on platform CUDA. Devices:\n",
      "2019-08-26 17:01:07.540374: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): GeForce GTX TITAN X, Compute Capability 5.2\n",
      "2019-08-26 17:01:07.541596: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1433] Found device 0 with properties: \n",
      "name: GeForce GTX TITAN X major: 5 minor: 2 memoryClockRate(GHz): 1.076\n",
      "pciBusID: 0000:0a:00.0\n",
      "totalMemory: 11.93GiB freeMemory: 11.82GiB\n",
      "2019-08-26 17:01:07.541657: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1512] Adding visible gpu devices: 0\n",
      "2019-08-26 17:01:07.543812: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:\n",
      "2019-08-26 17:01:07.543847: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0 \n",
      "2019-08-26 17:01:07.543864: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N \n",
      "2019-08-26 17:01:07.544842: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 11498 MB memory) -> physical GPU (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:0a:00.0, compute capability: 5.2)\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m dataloader.output_schema is compatible with model.schema\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m Initialized data generator. Running batches...\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.specs]\u001b[0m Example file for argument intervals_file already exists\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.specs]\u001b[0m Example file for argument fasta_file already exists\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.specs]\u001b[0m Example file for argument vcf_file already exists\u001b[0m\n",
      "Downloading https://zenodo.org/record/3374833/files/chr22utrVar1000gen_sorted.vcf.gz.tbi?download=1 to /data/ouga04b/ag_gagneur/home/karollus/5UTRModel/Collab/kipoi/5UTR_Model/downloaded/example_files/vcf_file_tbi\n",
      "8.19kB [00:00, 18.7kB/s]                                                        \n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.specs]\u001b[0m Example file for argument chr_order_file already exists\u001b[0m\n",
      "/data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/kipoiseq/dataloaders/sequence.py:88: FutureWarning: read_table is deprecated, use read_csv instead.\n",
      "  sep='\\t')\n",
      "/data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/kipoiseq/dataloaders/sequence.py:100: FutureWarning: read_table is deprecated, use read_csv instead.\n",
      "  sep='\\t')\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m Returned data schema correct\u001b[0m\n",
      "  0%|                                                    | 0/45 [00:00<?, ?it/s]2019-08-26 17:01:13.751553: I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library libcublas.so.10.0 locally\n",
      "100%|███████████████████████████████████████████| 45/45 [00:14<00:00,  5.67it/s]\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m predict_example done!\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.cli.main]\u001b[0m Successfully ran test_predict\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!kipoi test \"5UTR_Model\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using downloaded and verified file: /data/ouga04b/ag_gagneur/home/karollus/5UTRModel/Collab/kipoi/5UTR_Model/downloaded/model_files/weights/1bac703c830d6322dbace78721b1e405\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Colocations handled automatically by placer.\n",
      "WARNING:tensorflow:From /data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.\n",
      "WARNING:tensorflow:From /data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Use tf.cast instead.\n",
      "WARNING:tensorflow:From /data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/tensorflow/python/ops/math_grad.py:102: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Deprecated in favor of operator or tf.math.divide.\n"
     ]
    }
   ],
   "source": [
    "# Source model\n",
    "model = kipoi.get_model(\"5UTR_Model\", source=\"dir\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path of the vcf file\n",
    "vcf_path = \"TestFiles/stopLoss.vcf.gz\"\n",
    "\n",
    "# Path of the fasta file\n",
    "fasta_path = \"/s/genomes/human/hg19/fasta/hg19.fa\"\n",
    "# Set true if fasta has no chr prefix, false otherwise\n",
    "num_chr = False\n",
    "\n",
    "# Path of the bed file specifying human 5utr\n",
    "# If chr prefix is present, use:\n",
    "# gencodev19_5utr_sorted.bed, else:\n",
    "# gencodev19_5utr_sorted_noprefix.bed\n",
    "bed_path = \"TestFiles/gencodev19_5utr_sorted.bed\"\n",
    "id_column = 4\n",
    "\n",
    "# \n",
    "#chr_order_file = \"TestFiles/chrom_order_noprefix.txt\"\n",
    "\n",
    "# output file path\n",
    "output_file_path = \"patho.tsv\"\n",
    "\n",
    "# id map path\n",
    "id_map_path = \"TestFiles/hg19_idmap.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/kipoiseq/dataloaders/sequence.py:88: FutureWarning: read_table is deprecated, use read_csv instead.\n",
      "  sep='\\t')\n",
      "/data/ouga04b/ag_gagneur/home/karollus/.conda/envs/karollus-env/lib/python3.6/site-packages/kipoiseq/dataloaders/sequence.py:100: FutureWarning: read_table is deprecated, use read_csv instead.\n",
      "  sep='\\t')\n",
      "100%|██████████| 1/1 [00:04<00:00,  4.98s/it]\n"
     ]
    }
   ],
   "source": [
    "model.pipeline.predict_to_file(output_file_path, {\"intervals_file\":bed_path, \n",
    "                               \"fasta_file\":fasta_path,\n",
    "                               \"vcf_file\":vcf_path,\n",
    "                               \"id_column\":id_column,\n",
    "                               \"num_chr\":num_chr},\n",
    "                              batch_size=64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data as dataframe and merge with gene names\n",
    "df = pd.read_csv(output_file_path, sep=\"\\t\")\n",
    "df = df.rename(index=str, columns={\"metadata/chr\":\"chr\",\n",
    "          \"metadata/exon_positions\":\"exon_positions\",\n",
    "          \"metadata/id\":\"transcript_id\",\n",
    "          \"metadata/strand\":\"strand\",\n",
    "          \"metadata/variants\":\"variants\",\n",
    "          \"preds/mrl_fold_change\":\"mrl_fold_change\",\n",
    "          \"preds/shift_1\":\"shift_1\",\n",
    "          \"preds/shift_2\":\"shift_2\"}\n",
    ")\n",
    "\n",
    "# Id map\n",
    "df_map = pd.read_csv(id_map_path, sep=\"\\t\")\n",
    "df = df.merge(df_map, on=\"transcript_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>exon_positions</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>variants</th>\n",
       "      <th>mrl_fold_change</th>\n",
       "      <th>shift_1</th>\n",
       "      <th>shift_2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>68048839-68049619</td>\n",
       "      <td>ENST00000204961</td>\n",
       "      <td>+</td>\n",
       "      <td>chrX:68049525:T:['C']</td>\n",
       "      <td>-0.081899</td>\n",
       "      <td>0.106978</td>\n",
       "      <td>-0.115071</td>\n",
       "      <td>ENSG00000090776</td>\n",
       "      <td>EFNB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>159593490-159593608</td>\n",
       "      <td>ENST00000307738</td>\n",
       "      <td>+</td>\n",
       "      <td>chr4:159593534:A:['G']</td>\n",
       "      <td>-0.140904</td>\n",
       "      <td>-0.059372</td>\n",
       "      <td>0.193931</td>\n",
       "      <td>ENSG00000171503</td>\n",
       "      <td>ETFDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>21986683-21986723;21987940-21988565</td>\n",
       "      <td>ENST00000312841</td>\n",
       "      <td>-</td>\n",
       "      <td>chr8:21988118:T:['C']</td>\n",
       "      <td>0.102993</td>\n",
       "      <td>-0.070834</td>\n",
       "      <td>-0.041693</td>\n",
       "      <td>ENSG00000168453</td>\n",
       "      <td>HR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>29421944-29422327</td>\n",
       "      <td>ENST00000356175</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:29422055:A:['C']</td>\n",
       "      <td>-0.084574</td>\n",
       "      <td>-0.048338</td>\n",
       "      <td>0.237754</td>\n",
       "      <td>ENSG00000196712</td>\n",
       "      <td>NF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>29421944-29422327</td>\n",
       "      <td>ENST00000358273</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:29422055:A:['C']</td>\n",
       "      <td>-0.084574</td>\n",
       "      <td>-0.048338</td>\n",
       "      <td>0.237754</td>\n",
       "      <td>ENSG00000196712</td>\n",
       "      <td>NF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>21986683-21986723;21987940-21989381</td>\n",
       "      <td>ENST00000381418</td>\n",
       "      <td>-</td>\n",
       "      <td>chr8:21988118:T:['C']</td>\n",
       "      <td>0.004778</td>\n",
       "      <td>-0.007438</td>\n",
       "      <td>-0.002313</td>\n",
       "      <td>ENSG00000168453</td>\n",
       "      <td>HR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>29421994-29422327</td>\n",
       "      <td>ENST00000431387</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:29422055:A:['C']</td>\n",
       "      <td>-0.086739</td>\n",
       "      <td>-0.049234</td>\n",
       "      <td>0.256214</td>\n",
       "      <td>ENSG00000196712</td>\n",
       "      <td>NF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr4</td>\n",
       "      <td>159593276-159593608</td>\n",
       "      <td>ENST00000511912</td>\n",
       "      <td>+</td>\n",
       "      <td>chr4:159593534:A:['G']</td>\n",
       "      <td>-0.032477</td>\n",
       "      <td>-0.029689</td>\n",
       "      <td>-0.015868</td>\n",
       "      <td>ENSG00000171503</td>\n",
       "      <td>ETFDH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr                       exon_positions    transcript_id strand  \\\n",
       "0   chrX                    68048839-68049619  ENST00000204961      +   \n",
       "1   chr4                  159593490-159593608  ENST00000307738      +   \n",
       "2   chr8  21986683-21986723;21987940-21988565  ENST00000312841      -   \n",
       "3  chr17                    29421944-29422327  ENST00000356175      +   \n",
       "4  chr17                    29421944-29422327  ENST00000358273      +   \n",
       "5   chr8  21986683-21986723;21987940-21989381  ENST00000381418      -   \n",
       "6  chr17                    29421994-29422327  ENST00000431387      +   \n",
       "7   chr4                  159593276-159593608  ENST00000511912      +   \n",
       "\n",
       "                 variants  mrl_fold_change   shift_1   shift_2  \\\n",
       "0   chrX:68049525:T:['C']        -0.081899  0.106978 -0.115071   \n",
       "1  chr4:159593534:A:['G']        -0.140904 -0.059372  0.193931   \n",
       "2   chr8:21988118:T:['C']         0.102993 -0.070834 -0.041693   \n",
       "3  chr17:29422055:A:['C']        -0.084574 -0.048338  0.237754   \n",
       "4  chr17:29422055:A:['C']        -0.084574 -0.048338  0.237754   \n",
       "5   chr8:21988118:T:['C']         0.004778 -0.007438 -0.002313   \n",
       "6  chr17:29422055:A:['C']        -0.086739 -0.049234  0.256214   \n",
       "7  chr4:159593534:A:['G']        -0.032477 -0.029689 -0.015868   \n",
       "\n",
       "           gene_id gene_name  \n",
       "0  ENSG00000090776     EFNB1  \n",
       "1  ENSG00000171503     ETFDH  \n",
       "2  ENSG00000168453        HR  \n",
       "3  ENSG00000196712       NF1  \n",
       "4  ENSG00000196712       NF1  \n",
       "5  ENSG00000168453        HR  \n",
       "6  ENSG00000196712       NF1  \n",
       "7  ENSG00000171503     ETFDH  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
